framework/quota/src/org/apache/cloudstack/quota/dao/QuotaBalanceDaoImpl.java

```diff
@@ -0,0 +1,201 @@
+import java.util.Iterator;
+import java.util.List;
+
+import javax.ejb.Local;
+
+import org.springframework.stereotype.Component;
+import org.apache.cloudstack.quota.vo.QuotaBalanceVO;
+import org.apache.log4j.Logger;
+
+import com.cloud.exception.InvalidParameterValueException;
+import com.cloud.utils.db.Filter;
+import com.cloud.utils.db.GenericDaoBase;
+import com.cloud.utils.db.SearchCriteria;
+import com.cloud.utils.db.TransactionLegacy;
+
+@Component
+@Local(value = { QuotaBalanceDao.class })
+public class QuotaBalanceDaoImpl extends GenericDaoBase<QuotaBalanceVO, Long> implements QuotaBalanceDao {
+    private static final Logger s_logger = Logger.getLogger(QuotaBalanceDaoImpl.class.getName());
+
+    @SuppressWarnings("deprecation")
+    @Override
+    public QuotaBalanceVO findLastBalanceEntry(final long accountId, final long domainId, final Date beforeThis) {
+        final short opendb = TransactionLegacy.currentTxn().getDatabaseId();
+        TransactionLegacy.open(TransactionLegacy.USAGE_DB).close();
+        Filter filter = new Filter(QuotaBalanceVO.class, "updatedOn", false, 0L, 1L);
+        SearchCriteria<QuotaBalanceVO> sc = createSearchCriteria();
+        sc.addAnd("accountId", SearchCriteria.Op.EQ, accountId);
+        sc.addAnd("domainId", SearchCriteria.Op.EQ, domainId);
+        sc.addAnd("creditsId", SearchCriteria.Op.EQ, 0);
+        sc.addAnd("updatedOn", SearchCriteria.Op.LT, beforeThis);
+        List<QuotaBalanceVO> quotab = this.search(sc, filter);
+        TransactionLegacy.open(opendb).close();
+        return quotab.size() > 0 ? quotab.get(0) : null;
+    }
+
+    @SuppressWarnings("deprecation")
+    @Override
+    public QuotaBalanceVO findLaterBalanceEntry(final long accountId, final long domainId, final Date afterThis) {
+        final short opendb = TransactionLegacy.currentTxn().getDatabaseId();
+        TransactionLegacy.open(TransactionLegacy.USAGE_DB).close();
+        Filter filter = new Filter(QuotaBalanceVO.class, "updatedOn", true, 0L, 1L);
+        SearchCriteria<QuotaBalanceVO> sc = createSearchCriteria();
+        sc.addAnd("accountId", SearchCriteria.Op.EQ, accountId);
+        sc.addAnd("domainId", SearchCriteria.Op.EQ, domainId);
+        sc.addAnd("creditsId", SearchCriteria.Op.EQ, 0);
+        sc.addAnd("updatedOn", SearchCriteria.Op.GT, afterThis);
+        List<QuotaBalanceVO> quotab = this.search(sc, filter);
+        TransactionLegacy.open(opendb).close();
+        return quotab.size() > 0 ? quotab.get(0) : null;
+    }
+
+    @Override
+    public void saveQuotaBalance(final List<QuotaBalanceVO> credits) {
+        final short opendb = TransactionLegacy.currentTxn().getDatabaseId();
+        TransactionLegacy txn = TransactionLegacy.open(TransactionLegacy.USAGE_DB);
+        try {
+            for (QuotaBalanceVO credit : credits) {
+                persist(credit);
+            }
+        } finally {
+            txn.close();
+        }
+        TransactionLegacy.open(opendb).close();
```

Comment:
See previous comments regarding apparently needless open/close of a database transaction.